temporal-serviceclient/src/main/java/io/temporal/internal/AbstractBackoffThrottler.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.internal;
+
+import io.grpc.Status;
+import java.time.Duration;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+
+/**
+ * Used to throttle code execution in presence of failures using exponential backoff logic. The
+ * formula used to calculate the next sleep interval is:
+ *
+ * <p>
+ *
+ * <pre>
+ * min(pow(backoffCoefficient, failureCount - 1) * initialSleep, maxSleep);
+ * </pre>
+ */
+abstract class AbstractBackoffThrottler {
```

Comment:
This is a typical example where composition should be chosen over an inheritance.
You don't even need or use polymorphism here.